File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No82RemoveDuplicatesFromSortedListII.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import com.hazza.LeetCodeByJava.datastructure.ListNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-05
 * Time: 9:18 AM
 */
public class No82RemoveDuplicatesFromSortedListII {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return head;
        ListNode temp = new ListNode(0);
        temp.next = head;
        ListNode prev = temp;
        while (prev.next != null) {
            ListNode cur = prev.next;
            boolean duplicate = false;
            while (cur.next != null && cur.val == cur.next.val) {
                cur = cur.next;
                duplicate = true;
            }
            if (duplicate) prev.next = cur.next;
            else prev = prev.next;
        }

        return temp.next;
    }
}

File: src/main/python/com/hazza/LeetCodeByPython/linked_list/No82_Remove_Duplicates_from_Sorted_List_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-18
# https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
from src.main.python.com.hazza.LeetCodeByPython.data_structure.ListNode import makeListNode, isListNodeEqual, ListNode, \
    printListNode


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        temp = ListNode(0)
        temp.next = head
        pre = temp

        while pre.next:
            isDuplicate = False
            cur = pre.next
            while cur.next and cur.next.val == cur.val:
                cur = cur.next
                isDuplicate = True
            if isDuplicate:
                pre.next = cur.next
            else:
                pre = pre.next

        return temp.next


if __name__ == '__main__':
    solution = Solution()

    head = makeListNode([1, 2, 3, 3, 4, 4, 5])
    expected = makeListNode([1, 2, 5])
    res = solution.deleteDuplicates(head)
    printListNode(res)
    assert isListNodeEqual(res, expected)

    head = makeListNode([1, 1, 1, 2, 3])
    expected = makeListNode([2, 3])
    res = solution.deleteDuplicates(head)
    printListNode(res)
    assert isListNodeEqual(res, expected)
